### DISCUSSION TITLE
Retrieve an item with public role from a Hook

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#7601] [Archive]
Hi,
From a hook after the creation of the item, I would like to retrieve the item but to read it with a public role. 
My intention is to index the item but only with the public readable properties.

What I have tried so far is the following

```
const registerHook: HookRegisterFunction = ({ services }) => {
	return {
		'items.create': async function ({ collection, item: itemId, schema, accountability }) {
			const itemService = new services.ItemsService(collection, {
				accountability: {
					...accountability,
					role: null,
				}, schema
			})
			const item = await itemService.readOne(itemId);
		},
	};
};

export default registerHook;

```

In case I send the admin property to false for the accountability, I cannot read at all the object. ('You don't have permission to access this'). In case it is set to true, the role is ignored and I can read the entire object properties as an admin.

My collection is publicly readable.

Thanks for help!

**Comment by rijkvanzanten** [comment]
The public user would be `null` for `user`, and `role`, and `false` for `admin`, and `app`.

If you get a `You don't have permission to access this` with those flags, you might have to double check the configured read permissions for the public role, as it means that the item isn't allowed to be read, or doesn't exist.

**Reply by gkielwasser** [reply]
I have tried with this accountibility object and got the same permission denied issue. I have also double check that the collection I try to access has the public read access. I have also done a permissions check to be sure, directly querying the collection without access token the api.

```
accountability: {
	role: null,
	user: null,
	admin: false,
	app: false,
}
```

About the existence of the object in this context: I would say it exists because just updating the admin flag to true, makes the item to be found. 
```
accountability: {
	role: null,
	user: null,
	admin: true
	app: false,
}
```

Is the schema argument used also to check permissions or everything rely on the accountability object?

**Reply by rijkvanzanten** [reply]
Ahh good one. Just verified, and the current user's permissions are within schema as well. I'll open a new issue to move that to `accountability`, as that shouldn't' be under `schema` 👍🏻 

Try setting `schema.permissions` to `[]` or `null`

**Reply by rijkvanzanten** [reply]
ref https://github.com/directus/directus/issues/7636

**Reply by gkielwasser** [reply]
Ok, I have tried with the following.
```
schema.permissions = []
``` 
and 
```
accountability: {
	role: null,
	user: null,
	admin: false,
	app: false,
}
```
This one still give me ```You don't have permission to access this```

Whenether I try with a null permissions object, I got the following error 
``` Cannot read property 'filter' of null```

Looking this, it looks like it is a must to have a permission set without the admin property set to true, isn't it ?
What shall I put to mimic the public permission ?

Thanks

**Reply by rijkvanzanten** [reply]
Ahh right right. The easiest way would be to re-fetch the schema using the new accountability object you've created, and passing that on to the items service:

```js
const registerHook: HookRegisterFunction = ({ services, getSchema }) => {
	const publicAccountability = {
		role: null,
		user: null,
		admin: false,
		app: false,
	};

	return {
		'items.create': async function ({ collection, item: itemId }) {
			const schema = await getSchema({ accountability: publicAccountability });

			const itemService = new services.ItemsService(collection, {
				accountability: publicAccountability,
				schema,
			});

			const item = await itemService.readOne(itemId);
		},
	};
};

export default registerHook;
```

**Reply by gkielwasser** [reply]
Perfect !
This one works well,
Thank you !

**Reply by gkielwasser** [reply]
I cannot mark your last comment as the answer (who knows why), but it is the one 👍

**Reply by gkielwasser** [reply]
Hi @rijkvanzanten 

I'm not sure if I should open another discussion, but the issue is strictly related to this one!

The above solution was working well but it has stopped working, at least for Directus 9.0.1.

Is there somehting that has changed about permissions that would have invalidated the above solution ?

Anything about the "undefined" "permissions" array?

Thanks !

**Reply by azrikahar** [reply]
@gkielwasser assuming by "stopped working", you mean the above solution without any changes, then it's likely due to hooks being reworked back in 9.0.0-rc.101 as mentioned in the release note: https://github.com/directus/directus/releases/tag/v9.0.0-rc.101 👍

link to docs on the how to adapt your code to the new syntax: https://docs.directus.io/extensions/hooks/

---

With that said, here's an attempt to "translate" your code to the current syntax without change too much of the general structure:

_Note that the following code is untested & may require some tweaks_

```js
const registerHook: HookRegisterFunction = ({ action }, { services, getSchema }) => {
	const publicAccountability = {
		role: null,
		user: null,
		admin: false,
		app: false,
	};

        action('items.create', async function ({ key, collection }, context) {
	        const schema = await getSchema({ accountability: publicAccountability });
        
	        const itemService = new services.ItemsService(collection, {
		        accountability: publicAccountability,
		        schema,
	        });
        
	        const item = await itemService.readOne(key);
        });
};

export default registerHook;
```

**Reply by gkielwasser** [reply]
> @gkielwasser assuming by "stopped working", you mean the above solution without any changes, then it's likely due to hooks being reworked back in 9.0.0-rc.101 as mentioned in the release note: https://github.com/directus/directus/releases/tag/v9.0.0-rc.101 👍
> 
> link to docs on the how to adapt your code to the new syntax: https://docs.directus.io/extensions/hooks/
> 
> With that said, here's an attempt to "translate" your code to the current syntax without change too much of the general structure:
> 
> _Note that the following code is untested & may require some tweaks_
> 
> ```js
> const registerHook: HookRegisterFunction = ({ action }, { services, getSchema }) => {
> 	const publicAccountability = {
> 		role: null,
> 		user: null,
> 		admin: false,
> 		app: false,
> 	};
> 
>         action('items.create', async function ({ key, collection }, context) {
> 	        const schema = await getSchema({ accountability: publicAccountability });
>         
> 	        const itemService = new services.ItemsService(collection, {
> 		        accountability: publicAccountability,
> 		        schema,
> 	        });
>         
> 	        const item = await itemService.readOne(key);
>         });
> };
> 
> export default registerHook;
> ```

Ups @azrikahar, you are right, I didn't say what stopped working. I have already the new hooks strategy.

What is not working anymore is this part
```
const publicAccountability = {
		role: null,
		user: null,
		admin: false,
		app: false,
	};
const itemService = new services.ItemsService(collection, {
        accountability: publicAccountability,
	schema,
 });
```
It doesn't mimic anymore an itemService that would read with public permissions. It says it hasn't permissions to read items (On a publicly readable collection)

**Reply by azrikahar** [reply]
ah... just read up the hidden `Show 5 previous replies` above and I believe the reason it doesn't work now is because the move of permissions to accountability from schema, as mentioned by Rijk, was recently done.

The following code _should_ work, but I'm also not sure if there's any way to make it simpler:

```js
module.exports = function registerHook({ action }, { services, getSchema }) {
	const publicAccountability = {
		role: null,
		user: null,
		admin: false,
		app: false,
	};

	action('items.create', async function ({ key, collection }, context) {
		const schema = await getSchema();

                // get permissions for public role, which is "role equals null"
		const permissionsService = new services.PermissionsService({ schema });
		const permissions = await permissionsService.readByQuery({
			filter: {
				role: {
					_null: true,
				},
			},
		});

		const itemService = new services.ItemsService(collection, {
			accountability: { ...publicAccountability, permissions },
			schema,
		});

		const item = await itemService.readOne(key);
	});
};
```

basically we'll need to attach the `permissions` array to your accountability now, which wasn't required in the past.

**Reply by gkielwasser** [reply]
Perfect ! I suspected I had to retrieve those public permissions now.

For now I will retrieve those permissions "on demand"

Thanks !

**Reply by rijkvanzanten** [reply]
I shouted this out as part of the release notes, as I remembered this coming up in a discussion (this one!) 🙂 

<img width="1081" alt="CleanShot 2021-11-12 at 15 46 12@2x" src="https://user-images.githubusercontent.com/9141017/141532375-401910f9-338c-4bc4-bcec-1dffb6045c68.png">

**Reply by azrikahar** [reply]
Yea forgot to link to that in my previous reply 👍 Just wondering is the sample approach correct? In case there's a much simpler approach then what I came up with.

**Reply by rijkvanzanten** [reply]
@azrikahar Your approach should work just fine 👍🏻
